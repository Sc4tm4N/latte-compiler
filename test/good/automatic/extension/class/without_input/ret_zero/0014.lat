class a {

  void print() {
    printString("i am a, my a is:");
    printInt(a());
  }

  int a() {
    return 0;
  }

  int ref() {
    return 1000;
  }
}

class b extends a {

  void print() {
    printString("i am b, my a is:");
    printInt(a());
  }

  int a() {
    return b();
  }

  int b() {
    return ref();
  }
}

class c extends b {

  void print() {
    printString("i am c, my a is:");
    printInt(a());
  }

  int a() {
    return b() * self.c();
  }

  int c() {
    return 2 + c();
  }

  int ref() {
    return 666;
  }
}

int c() {
  return 20;
}

int main() {
  a a = new a;
  a.print();
  printInt(a.a());
  printInt(a.ref());

  b b = new b;
  b.print();
  printInt(b.a());
  printInt(b.b());
  printInt(b.ref());

  c c = new c;
  c.print();
  printInt(c.a());
  printInt(c.b());
  printInt(c.c());
  printInt(c.ref());

  a ab = new b;
  ab.print();
  printInt(ab.a());
  printInt(ab.ref());

  a ac = new c;
  ac.print();
  printInt(ac.a());
  printInt(ac.ref());

  b bc = new c;
  bc.print();
  printInt(bc.a());
  printInt(bc.b());
  printInt(bc.ref());

  return 0;
}
