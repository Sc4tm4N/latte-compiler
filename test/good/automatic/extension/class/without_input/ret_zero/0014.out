i am a, my a is:
0
0
1000
i am b, my a is:
1000
1000
1000
1000
i am c, my a is:
14652
14652
666
22
666
i am b, my a is:
1000
1000
1000
i am c, my a is:
14652
14652
666
i am c, my a is:
14652
14652
666
666
